caffe2/opt/optimize_ideep.cc

```diff
@@ -2,42 +2,404 @@
+    repr::NNGraph::NodeRef sumInputX
+      = (sumInputs[0] == convOutput ? sumInputs[1] : sumInputs[0]);
+    assert(sumInputX != nullptr && "Invalid sum inputs");
+
+    auto newOutputName = repr::nn::get<repr::Tensor>(sumInputX)->getName();
+    auto newOutputTensor = util::make_unique<repr::Tensor>(newOutputName);
+    auto newOutput = nn->dataFlow.createNode(
+        unique_dyn_cast<repr::NeuralNetData>(newOutputTensor));
+
+    auto sumOutput = repr::nn::getOutputs(sumNode).front();
+    nn->dataFlow.replaceNode(sumOutput, newOutput);
+
+    // 2 means FUSION_CONV_SUM
+    resetConvForFusion(convNode, 2);
+    nn->dataFlow.createEdge(sumInputX, convNode);
+    nn->dataFlow.createEdge(convNode, newOutput);
+
+    nn->dataFlow.deleteNode(sumNode);
+    nn->dataFlow.deleteNode(sumOutput);
+    nn->dataFlow.deleteNode(convOutput);
+  }
+}
+
+void fuseActivationForIdeep(repr::NNModule* nn) {
+  // Conv+Relu fusion
+  auto should_fuse = shouldFuseConv;
+  auto postprocess = std::bind(resetConvForFusion, std::placeholders::_1, 1);
   fuseActivation<repr::Conv, repr::Relu>(nn, should_fuse, postprocess);
 }
 
+void enforceFusionInplaceForIdeep(repr::NNModule* nn) {
+  // For fusions of Conv+Sum or Conv+Sum+ReLU, the last input and output must
+  // be inplaced. To enforce inplace, here to re-check whole graph and correct
+  // the ConvFusion Ops.
+  for (auto node_pair : repr::nn::dataIterator<repr::Conv>(nn->dataFlow)) {
+    repr::NNGraph::NodeRef convNode;
+    repr::Conv* conv;
+    std::tie(conv, convNode) = node_pair;
+
+    if (!isOnIdeepDevice(*conv)) {
+      LOG(WARNING) << "Not a IDEEP operator";
+      continue;
+    }
+
+    const auto* op = getOpDef(*conv);
+    if (op->type() != "ConvFusion") {
+      continue;
+    }
+
+    bool enforce_inplace = false;
+    for (const auto& arg : op->arg()) {
+      // Only check FUSION_SUM & FUSION_SUM_RELU
+      if (arg.name() == "fusion_type" && (arg.i() == 2 || arg.i() == 3)) {
+        enforce_inplace = true;
+        break;
+      }
+    }
+
+    if (!enforce_inplace) {
+      continue;
+    }
+
+    auto convInput = repr::nn::getInputs(convNode).back();
+    auto inputName = repr::nn::get<repr::Tensor>(convInput)->getName();
+    auto convOutput = repr::nn::getOutputs(convNode).front();
+    auto outputName = repr::nn::get<repr::Tensor>(convOutput)->getName();
+    if (inputName == outputName) {
+      continue;
+    }
+
+    auto consumer = repr::nn::getConsumers(convInput).back();
+    if (consumer != convNode) {
+      LOG(ERROR) << "Can not enforce to inplace for fusion";
+      return;
+    }
+
+    auto newOutputTensor = util::make_unique<repr::Tensor>(inputName);
+    auto newOutput = nn->dataFlow.createNode(
+        unique_dyn_cast<repr::NeuralNetData>(newOutputTensor));
+    nn->dataFlow.replaceNode(convOutput, newOutput);
+
+    nn->dataFlow.deleteNode(convOutput);
+  }
+}
+
+void preConvertFiltersFormat(repr::NNModule* nn, caffe2::Workspace* ws) {
```

Comment:
What does this do?

Reply:
Convert filter formats to optimized one just one time before runing.